specs/agents/metrics-health.md

```diff
@@ -0,0 +1,98 @@
+| 0s       | 7           | 7                         |
+| 30s      | 10          | 3                         |
+| 1m       | 10          | 0 (not sent)              |
+| 1m30s    | 12          | 2                         |
+
+
+### Event Count Metrics
+
+Agents send their data to the Intake API of the APM-server in the form of a stream of events (e.g. transactions, spans, metricsets...).
+
+The following metrics counting these events SHOULD be exposed:
+ - `agent.events.total`: `COUNTER` of events either submitted to the internal reporting queue or dropped upon submit attempt
+ - `agent.events.dropped.queue`: `COUNTER` of events where the attempt to add it to the reporting queue failed (e.g. due to a full queue)
+ - `agent.events.dropped.error`: `COUNTER` of events dropped after they previously have been successfully added to the queue.
+
+If an agent supports these metrics, the metrics MUST be captured for all events which would reach the APM-Server in an ideal, failure-free world.
+Therefore an event MUST be taken into account for these metrics if an attempt is made to submit it to the internal reporting queue, regardless of the success.
+Events dropped before this attempt (e.g. due to [sampling](tracing-sampling.md) or [transaction_max_spans](handling-huge-traces/tracing-spans-limit.md)) are NOT counted.
+
+The value of `agent.events.dropped.error` is an upper bound for the actual number of dropped events after entering the queue.
+For example, when an request to the Intake API fails without a response, all events within this request MUST be considered as failed.
+In reality, it is possible that e.g. half of the data was actually successfully received and forwarded by the APM server.
+When the APM server responds with an HTTP error code, the number of dropped events SHOULD be computed by subtracting the `accepted` field from the response body from the total number of inflight events of this request.
+
+The `agent.events.total` metric MUST have the label `eventType`. The possible values for `eventType` are the lowercase names of the events according to the [spec](https://github.com/elastic/apm-server/tree/main/docs/spec/v2) (e.g. `transaction`, `metricset`) or the value `other`.
+The `agent.events.dropped.*` metrics MUST NOT have any labels.
+
+Agents SHOULD attempt to assign the appropriate label value based on the counted event. If this would impose significant implementation overhead, the value `other` MUST be used instead.
+
+All event count metrics can be disabled via the `disable_metrics` configuration.
+
+### Event Queue Metrics
+
+The following metrics SHOULD be recorded by agents regarding the state of the internal event reporting queue:
+
+ - `agent.events.queue.min_size.pct`: The smallest queue usage in percent (range [0-1]) observed since this metric was reported the last time
+ - `agent.events.queue.max_size.pct`: The biggest queue usage in percent (range [0-1]) observed since this metric was reported the last time
```

Comment:
> Do you have an idea in mind of how these metrics should be calculated? If we were to report just a single current utilisation metric it would be clear: just calculate it on the usual metrics interval.

Yes, this is how I've done it for the Java agent:
I have variables storing the current `minQueueSize` and `maxQueueSize`. Whenever the queue size changes (event is added or removed), these variables get updated with an [atomic min/max](https://github.com/JonasKunz/apm-agent-java/blob/93c3e6f765e3af41094a06bc71831b97d4877550/apm-agent-core/src/main/java/co/elastic/apm/agent/util/AtomicDouble.java#L52). In addition when an event gets dropped, I set `maxQueueSize` to `1.0`.

When the metrics are reported, the current values of `minQueueSize` and `maxQueueSize` are reported and the variables are reset to the current queue size.

> This seems a bit like [Pressure Stall Information](https://facebookmicrosites.github.io/psi/docs/overview). Here it would be more like the amount of time (over the reporting period) where the queue was full.

Yeah seems similar. However, if I understand it correctly, PSI aims to monitor induced latency due to a contended resource. We however do not induce any latency when reporting events to the queue, as we drop them instead of waiting due to a full queue. Monitoring the time the queue is full would be different to PSI: E.g. if your resource is full for 1 second, but no new task arrives, your PSI is still zero.

Reply:
> Yes, this is how I've done it for the Java agent:
I have variables storing the current minQueueSize and maxQueueSize. Whenever the queue size changes (event is added or removed), these variables get updated with an [atomic min/max](https://github.com/JonasKunz/apm-agent-java/blob/93c3e6f765e3af41094a06bc71831b97d4877550/apm-agent-core/src/main/java/co/elastic/apm/agent/util/AtomicDouble.java#L52). In addition when an event gets dropped, I set maxQueueSize to 1.0.
>
> When the metrics are reported, the current values of minQueueSize and maxQueueSize are reported and the variables are reset to the current queue size.

Thanks, makes sense. Maybe that brief description to the spec, to assist other agent devs in their implementation?

> Monitoring the time the queue is full would be different to PSI: E.g. if your resource is full for 1 second, but no new task arrives, your PSI is still zero.

Good point!

Reply:
> Whenever the queue size changes (event is added or removed), these variables get updated with an [atomic min/max](https://github.com/JonasKunz/apm-agent-java/blob/93c3e6f765e3af41094a06bc71831b97d4877550/apm-agent-core/src/main/java/co/elastic/apm/agent/util/AtomicDouble.java#L52).

Not necessarily a blocker but I'm a bit worried that this will induce latency overhead when a lot of threads are contending to add events to the queue. The disruptor queue we're using in the Java agent is carefully designed to minimize that overhead. Using a data structure that's designed around CAS loops to count min/max queue size could therefore add overhead when there's contention.